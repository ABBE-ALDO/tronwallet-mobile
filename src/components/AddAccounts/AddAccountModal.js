import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import MixPanel from 'react-native-mixpanel'

import tl from '../../utils/i18n'
import { logSentry } from '../../utils/sentryUtils'
import { getAccountsCreatorStore } from '../../lib/secrets'

import Modal from '../Modal'
import LoadingButtonGradient from '../Buttons/LoadingButtonGradient'
import { View } from '../Utils'
import { Colors } from '../DesignSystem'

import ModalHeader from './ModalHeader'
import AccountNameInput from './AccountNameInput'
import AccountTypeSelector from './AccountTypeSelector'

class AddAccountModal extends PureComponent {
  static propTypes = {
    visible: PropTypes.bool,
    closeModal: PropTypes.func.isRequired,
    onCreated: PropTypes.func,
    oneSignalId: PropTypes.string.isRequired,
    pin: PropTypes.string.isRequired,
    totalAccounts: PropTypes.number.isRequired
  }

  static defaultProps = {
    onCreated: () => {},
    visible: false
  }

  state = {
    accountName: '',
    accountNameInvalid: false,
    key: null,
    accountData: {}
  }

  _changeAccountName = (accountName, error) =>
    this.setState({ accountName, accountNameInvalid: !!error })

  _onChangeData = (data) => {
    const { key, accountData } = this.state
    this.setState({ accountData: { ...accountData, [key]: data } })
  }

  _onSelectType = (key) => this.setState({ key })

  _addNewAccount = async () => {
    this.setState({ creatingNewAccount: true })

    try {
      const { accountName, key, accountData } = this.state

      const accountsCreatorStore = await getAccountsCreatorStore()

      const data = {
        accountName,
        ...accountData[key]
      }

      const newAccount = accountsCreatorStore.addAccountByMode(data, key)

      if (newAccount) {
        return
      }

      this.props.onCreated()

      MixPanel.trackWithProperties('Account Operation', { type: 'Create new account' })
    } catch (error) {
      logSentry(error, 'Error creating new Account')
    } finally {
      this.setState({ creatingNewAccount: false })
    }
  }

  _disableAddAccount = () => {
    const { accountNameInvalid, key, accountData } = this.state
    let { valid } = accountData[key] || {}

    if (valid === undefined) {
      valid = key === 'default'
    }

    return accountNameInvalid || !valid
  }

  render () {
    const {
      visible,
      closeModal,
      pin,
      totalAccounts
    } = this.props

    const { creatingNewAccount } = this.state

    const addAccountDisabled = this._disableAddAccount()

    return (
      <Modal
        modalOpened={visible}
        closeModal={closeModal}
        animationType='fade'
        transparent
      >
        <View flex={1} justify='center' background={Colors.shadowBackground} padding={15}>
          <View borderRadius={5} background={Colors.dusk} padding={15}>
            <ModalHeader
              title={tl.t('newAccount.title')}
              onClose={closeModal}
            />
            <AccountNameInput
              pin={pin}
              totalAccounts={totalAccounts}
              onChangeText={this._changeAccountName}
            />
            <AccountTypeSelector
              onSelectType={this._onSelectType}
              onChangeData={this._onChangeData}
            />
            <LoadingButtonGradient
              text={tl.t(`import.button.create`)}
              onPress={this._addNewAccount}
              disabled={addAccountDisabled}
              loading={creatingNewAccount}
            />
          </View>
        </View>
      </Modal>
    )
  }
}

export default AddAccountModal
