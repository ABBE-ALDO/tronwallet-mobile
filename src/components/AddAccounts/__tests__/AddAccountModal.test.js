import React from 'react'
import { shallow } from 'enzyme'

import AddAccountModal from '../AddAccountModal'

jest.mock('../../../utils/i18n')

describe('Add account Modal component', () => {
  const setup = propOverrides => {
    const props = Object.assign({
      pin: '123456',
      totalAccounts: 1
    }, propOverrides)

    const wrapper = shallow(
      <AddAccountModal {...props} />
    )

    return {
      props,
      wrapper
    }
  }

  test('Should match when render component', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })
})
