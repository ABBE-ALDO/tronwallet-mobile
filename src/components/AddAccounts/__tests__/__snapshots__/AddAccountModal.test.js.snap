// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add account Modal component Should match when render component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddAccountModal
    onCreated={[Function]}
    pin="123456"
    totalAccounts={1}
    visible={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "animationType": "fade",
      "children": <Styled(View)
        align="stretch"
        background="rgba(0,0,0,0.5)"
        flex={1}
        justify="center"
        padding={15}
      >
        <Styled(View)
          align="stretch"
          background="#3f415d"
          borderRadius={5}
          justify="flex-start"
          padding={15}
        >
          <ModalHeader
            onClose={[Function]}
            title="Account Creation"
          />
          <AccountNameInput
            onChangeText={[Function]}
            pin="123456"
            totalAccounts={1}
          />
          <AccountTypeSelector />
          <ButtonGradient
            disabled={false}
            font="medium"
            leftRadius={4}
            onPress={[Function]}
            rightRadius={4}
            size="large"
            text="CREATE"
          />
        </Styled(View)>
      </Styled(View)>,
      "closeModal": undefined,
      "modalOpened": false,
      "transparent": true,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "align": "stretch",
        "background": "rgba(0,0,0,0.5)",
        "children": <Styled(View)
          align="stretch"
          background="#3f415d"
          borderRadius={5}
          justify="flex-start"
          padding={15}
        >
          <ModalHeader
            onClose={[Function]}
            title="Account Creation"
          />
          <AccountNameInput
            onChangeText={[Function]}
            pin="123456"
            totalAccounts={1}
          />
          <AccountTypeSelector />
          <ButtonGradient
            disabled={false}
            font="medium"
            leftRadius={4}
            onPress={[Function]}
            rightRadius={4}
            size="large"
            text="CREATE"
          />
        </Styled(View)>,
        "flex": 1,
        "justify": "center",
        "padding": 15,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "align": "stretch",
          "background": "#3f415d",
          "borderRadius": 5,
          "children": Array [
            <ModalHeader
              onClose={[Function]}
              title="Account Creation"
            />,
            <AccountNameInput
              onChangeText={[Function]}
              pin="123456"
              totalAccounts={1}
            />,
            <AccountTypeSelector />,
            <ButtonGradient
              disabled={false}
              font="medium"
              leftRadius={4}
              onPress={[Function]}
              rightRadius={4}
              size="large"
              text="CREATE"
            />,
          ],
          "justify": "flex-start",
          "padding": 15,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "onClose": [Function],
              "title": "Account Creation",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onChangeText": [Function],
              "pin": "123456",
              "totalAccounts": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": false,
              "font": "medium",
              "leftRadius": 4,
              "onPress": [Function],
              "rightRadius": 4,
              "size": "large",
              "text": "CREATE",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "animationType": "fade",
        "children": <Styled(View)
          align="stretch"
          background="rgba(0,0,0,0.5)"
          flex={1}
          justify="center"
          padding={15}
        >
          <Styled(View)
            align="stretch"
            background="#3f415d"
            borderRadius={5}
            justify="flex-start"
            padding={15}
          >
            <ModalHeader
              onClose={[Function]}
              title="Account Creation"
            />
            <AccountNameInput
              onChangeText={[Function]}
              pin="123456"
              totalAccounts={1}
            />
            <AccountTypeSelector />
            <ButtonGradient
              disabled={false}
              font="medium"
              leftRadius={4}
              onPress={[Function]}
              rightRadius={4}
              size="large"
              text="CREATE"
            />
          </Styled(View)>
        </Styled(View)>,
        "closeModal": undefined,
        "modalOpened": false,
        "transparent": true,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "align": "stretch",
          "background": "rgba(0,0,0,0.5)",
          "children": <Styled(View)
            align="stretch"
            background="#3f415d"
            borderRadius={5}
            justify="flex-start"
            padding={15}
          >
            <ModalHeader
              onClose={[Function]}
              title="Account Creation"
            />
            <AccountNameInput
              onChangeText={[Function]}
              pin="123456"
              totalAccounts={1}
            />
            <AccountTypeSelector />
            <ButtonGradient
              disabled={false}
              font="medium"
              leftRadius={4}
              onPress={[Function]}
              rightRadius={4}
              size="large"
              text="CREATE"
            />
          </Styled(View)>,
          "flex": 1,
          "justify": "center",
          "padding": 15,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "align": "stretch",
            "background": "#3f415d",
            "borderRadius": 5,
            "children": Array [
              <ModalHeader
                onClose={[Function]}
                title="Account Creation"
              />,
              <AccountNameInput
                onChangeText={[Function]}
                pin="123456"
                totalAccounts={1}
              />,
              <AccountTypeSelector />,
              <ButtonGradient
                disabled={false}
                font="medium"
                leftRadius={4}
                onPress={[Function]}
                rightRadius={4}
                size="large"
                text="CREATE"
              />,
            ],
            "justify": "flex-start",
            "padding": 15,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "onClose": [Function],
                "title": "Account Creation",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "onChangeText": [Function],
                "pin": "123456",
                "totalAccounts": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": false,
                "font": "medium",
                "leftRadius": 4,
                "onPress": [Function],
                "rightRadius": 4,
                "size": "large",
                "text": "CREATE",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenThreeAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
